src/main/java/org/truffleruby/language/TruffleBootNodes.java

```diff
@@ -94,23 +96,25 @@ public DynamicObject forceContext() {
         @Specialization
         public int main(VirtualFrame frame,
                 @Cached("create()") IndirectCallNode callNode,
-                @Cached("new()") SnippetNode findSFileSnippetNode,
-                @Cached("new()") SnippetNode checkSyntaxSnippetNode) {
+                @Cached("createOnSelf()") CallDispatchHeadNode findSFile,
+                @Cached("createOnSelf()") CallDispatchHeadNode checkSyntax,
+                @Cached("create()") StringNodes.MakeStringNode makeStringNode) {
 
-            setArgvGlobals();
+            setArgvGlobals(makeStringNode);
 
-            Source source = loadMainSourceSettingDollarZero(frame, findSFileSnippetNode);
+            Source source = loadMainSourceSettingDollarZero(frame, findSFile, makeStringNode);
```

Comment:
When using `createOnSelf()`, you can pass `null` for the `frame`. Therefore the whole specialization can be `@TruffleBoundary` which seems safer here given the amount of code and branches.